### ISSUE TITLE
Backing up 36TB via NFS or SFTP. 

### ISSUE THREAD

**Issue by todd1794** [restic/restic#1088]
## Output of `restic version`
restic 0.6.0 (v0.6.0-0-g820c88ea)
compiled with go1.8.3 on linux/amd64
restic 0.6.1
compiled with go1.8.3 on linux/amd64
restic 0.7.0
compiled with go1.8.3 on linux/amd64
restic version (The version that I changed parms on)
restic 0.7.0 (v0.7.0-12-g637de01-dirty)
compiled with go1.8.3 on linux/amd64

## How did you start restic exactly? (Include the complete command line)
restic -p /data/restic/jack_archive_key -r /data/restic/jack_jobs backup -e /data/restic/jack_archive_excludes /jack/jobs/

## What backend/server/service did you use?
backup local (VIA NFS)
backup SFTP

## Expected behavior
Faster Scan and Actual Transfer Speed

## Actual behavior
Slow Scan and transfer that would not finish for 38 days. 

## Steps to reproduce the behavior
Run with the defaults that don't work so well with 36 TB via SFTP or NFS.

## Do you have any idea what may have caused this?
The defaults which are in general AWESOME. Needed tuning to backup larger amounts of data and number of files.

Once I changed 
maxConcurrentBlobs = 32 -> 196
maxConcurrency     = 10 -> 60
in archiver.go

And switched to .7.0 My scan time went from 9 hours to 4 hours. My first run when from 38 days (according to ETA and transfer rates) To 60 hours! Which for a first run is aok with me. (Still running) 

So the reason for the issue is what else could I change to make it even faster? I still have headroom on boths side (backup server and main server). I am getting around 145 megabytes/second right now. With a multi-threaded rsync I get around 450 megabytes/second. I have a feeling restic can beat rsync and use less resources. Thank you for creating this project! I have been using it home since 0.3.2 very well done!

**Comment by fd0** [comment]
Wow, 36TB! Thanks a lot for leaving an issue, I'd always wondered how restic behaves with such a huge amount of data... Can you elaborate a bit on how you store the data? Multiple discs directly, a RAID, NAS?

During the scanning phase, all files/dirs are walked. Usually (on a single disk/ssd) this won't be much faster if done concurrently, but the code will be much more complex. So at the moment this is just a single thread traversing the target dirs.

It's interesting that increasing the concurrency helps the runtime for you, that's (at least for me) unexpected. The problem usually is that restic uses too much concurrency for a single disc, because of that I'm currently reworking the archiver to not read files concurrently.

I think what may help much better is creating much larger pack files. restic splits all files/dirs into blobs (that's where the deduplication takes place) and then combines the blobs into pack files. The blobs are chosen so that the pack files are between 4 and 16 MiB in size. That's a limit that I'd increase, you can find it here: https://github.com/restic/restic/blob/master/src/restic/repository/packer_manager.go#L42-L44

Set the minimum pack file size to 128MiB and the max size to like 512MiB and report back what happens :)

**Comment by todd1794** [comment]
The 36TB is on a 50TB volume that is across 24 HUA723030ALA640 drives. Across two ARC-1882 controllers and has 64 GiB of RAM. It is a pretty old system 2011. But it performs well enough. The backup system has 24 ST32000641AS drives on one LSI Logic / Symbios Logic SAS2008 PCI-Express Fusion-MPT SAS-2 [Falcon] controller. mdadm raid 6 and 16 GiB of RAM.

So on the backup system 128MiB min and 512MiB max will not even make a repo. (I did not wait longer than 15 minutes. So I took the numbers down to 48MiB min and 128MiB max. The repo init's now. But the scan looks like this... [12:21] 1337 directories, 11519 files, 2.046 GiB. Usually at the 12 minute mark there is at least 60000 files and 100 GiB for this repo. I am going to try SFTP now.

**Comment by todd1794** [comment]
From the file server pushing to the backup server with 128MiB min and 512MiB max
I get this:
root@jack:~# restic -r sftp:root@192.168.1.105:/data/restic/jack_install_sftp_from_jack1 backup /data/install
enter password for repository: 
scan [/data/install]
scanned 98083 directories, 1648196 files in 7:06
panic: Write: write /tmp/restic-temp-pack-702573220: no space left on device 0 errors  ETA 57:53 

goroutine 40278 [running]:
restic/archiver.(*Archiver).saveChunk(0xc4222f4370, 0xc22520, 0xc4200123e8, 0x0, 0x5e85e, 0x1, 0xc4273ce000, 0x5e85e, 0x80000, 0xc4200b02c0, ...)
        src/restic/archiver/archiver.go:165 +0x395
created by restic/archiver.(*Archiver).SaveFile
        src/restic/archiver/archiver.go:239 +0x3a4
panic: Write: write /tmp/restic-temp-pack-436506610: no space left on device

goroutine 40015 [running]:
restic/archiver.(*Archiver).saveChunk(0xc4222f4370, 0xc22520, 0xc4200123e8, 0x52deb5, 0xf8df0, 0x1857d764a00000, 0xc43b16e000, 0xf8df0, 0xfa000, 0xc4200b02c0, ...)
        src/restic/archiver/archiver.go:165 +0x395
created by restic/archiver.(*Archiver).SaveFile
        src/restic/archiver/archiver.go:239 +0x3a4

**Comment by fd0** [comment]
Hm, the scan phase should not be affected by the larger pack files at all, that's odd. Not creating a repository is also very odd, that isn't affected at all.

The error is caused by `/tmp` filling up with unfinished pack files. You need to reduce the max number of packers (https://github.com/restic/restic/blob/master/src/restic/repository/packer_manager.go#L42-L44), otherwise restic uses up to `maxPackSize` * `maxPackers` bytes of space. You can also specify a temp directory to use with the environment variable `TMPDIR`.

I'd recommend using 128 and 512 MiB for the packers, and 10 `maxPackFiles`. Let's see how that works.

**Comment by todd1794** [comment]
I introduced a new set of servers. And they are working quite well with the 128 and 512 MiB with 100 maxPackFiles. But the limiting factor is now network. (The backup server only has a 1 gigabit connection) The primary server is VERY new installed 2017 in April. 64 GB of RAM
Status in progress:

[07:00:46][eastwood][~]# restic -r sftp:root@almodovar:/data/restic/eastwood backup /data/bigstore
enter password for repository: 
scan [/data/bigstore]
scanned 154493 directories, 3187285 files in 0:07
[17:41:23] 48.74%  100.443 MiB/s  6.100 TiB / 12.516 TiB  1571982 / 3341778 items  0 errors  ETA 18:36:23 


The other set of servers I have no stability with now. It would be help to have the parms that I am  changing appear in restic version. Is that an easy option? For example:
restic version
restic 0.7.0 (v0.7.0-16-g04d27ac-dirty maxConcurrentBlobs = 196 maxConcurrency = 60 in archiver.go const minPackSize = 64 * 1024 * 1024 const maxPackSize = 256 * 1024 * 1024 const maxPackers = 10 in packer_manager.go)
compiled with go1.8.3 on linux/amd64

But really the dreamworld scenario is run: 
restic testenvironment serving-host backup-host 
have that run for a couple of hours and find all of the best tuning options between those two hosts.

Either way I am more than happy to plug in new numbers and test things for you. I do actually need a working backup as well. And I found the first set of issues the super slow scan was just the backup server being overloaded. So I am starting over with the tests for those servers.

New error...

root@jack:~# export TMPDIR=/tmpSSD/tmp/restic
root@jack:~# restic -r sftp:root@192.168.1.105:/data/restic/jack_install_sftp_from_jack1 backup /data/install
enter password for repository: 
scan [/data/install]
scanned 98083 directories, 1648196 files in 11:12
panic: OpenFile: sftp: "Failure" (SSH_FX_FAILURE)9 GiB  1205048 / 1746279 items  0 errors  ETA 2:12:39 

goroutine 1172018 [running]:
restic/archiver.(*Archiver).saveChunk(0xc420315040, 0xc22520, 0xc4200123e8, 0x48dccc9, 0x8a7b2, 0x372a97900000, 0xc4431e6000, 0x8a7b2, 0x18a000, 0xc4222be0b0, ...)
        src/restic/archiver/archiver.go:165 +0x395
created by restic/archiver.(*Archiver).SaveFile
        src/restic/archiver/archiver.go:239 +0x3a4

That is with maxConcurrentBlobs = 196 maxConcurrency = 60 in archiver.go const minPackSize = 64 * 1024 * 1024 const maxPackSize = 256 * 1024 * 1024 const maxPackers = 10 in packer_manager.go

hmm... Anymore ideas?

Thanks

Todd

**Comment by fd0** [comment]
Hm, no, not at the moment. That's a generic 'an error occurred' in sftp, no details. hm.

**Comment by todd1794** [comment]
When maxPackSize has enough RAM in tmpfs is works quite nice! This is minPackSize = 128 * 1024 * 1024 const maxPackSize = 512 * 1024 * 1024 const maxPackers = 200. The limiting factor is the network speed on the backup node for this one.
 
First run:
restic -r sftp:root@almodovar:/data/restic/eastwood backup /data/bigstore
enter password for repository: 
scan [/data/bigstore]
scanned 154493 directories, 3187285 files in 0:07
[33:42:35] 100.00%  108.150 MiB/s  12.517 TiB / 12.516 TiB  3341781 / 3341778 items  0 errors  ETA 0:00 
duration: 33:42:35, 108.15MiB/s
snapshot 8c636c8b saved

Second run:
restic -r sftp:root@almodovar:/data/restic/eastwood backup /data/bigstore
enter password for repository: 
using parent snapshot 8c636c8b
scan [/data/bigstore]
scanned 154550 directories, 3187852 files in 0:14
[43:27] 100.00%  4.917 GiB/s  12.518 TiB / 12.518 TiB  3342404 / 3342402 items  0 errors  ETA 0:00 
duration: 43:27, 5034.38MiB/s
snapshot 083f3b72 saved

**Comment by todd1794** [comment]
A day of snapshots!

while true;do date;restic -p restic.key -r sftp:root@almodovar:/data/restic/eastwood backup /data/bigstore;sleep 21600;done                                                                                                                          
Mon Jul 10 13:23:47 PDT 2017
using parent snapshot 083f3b72
scan [/data/bigstore]
scanned 154552 directories, 3187880 files in 0:07
[34:14] 100.00%  6.241 GiB/s  12.518 TiB / 12.518 TiB  3342433 / 3342432 items  0 errors  ETA 0:00 
duration: 34:14, 6390.26MiB/s
snapshot 2412600f saved
Mon Jul 10 19:58:34 PDT 2017
using parent snapshot 2412600f
scan [/data/bigstore]
scanned 154555 directories, 3187902 files in 0:07
[33:49] 100.00%  6.318 GiB/s  12.519 TiB / 12.519 TiB  3342457 / 3342457 items  0 errors  ETA 0:00 
duration: 33:49, 6469.22MiB/s
snapshot 06315132 saved
Tue Jul 11 02:32:55 PDT 2017
using parent snapshot 06315132
scan [/data/bigstore]
scanned 154555 directories, 3187904 files in 0:07
[33:46] 100.00%  6.327 GiB/s  12.519 TiB / 12.519 TiB  3342459 / 3342459 items  0 errors  ETA 0:00 
duration: 33:46, 6476.96MiB/s
snapshot 09068f46 saved
Tue Jul 11 09:07:12 PDT 2017
using parent snapshot 09068f46
scan [/data/bigstore]
scanned 154555 directories, 3187904 files in 0:07
[33:48] 100.00%  6.317 GiB/s  12.519 TiB / 12.519 TiB  3342459 / 3342459 items  0 errors  ETA 0:01 
duration: 33:48, 6470.57MiB/s
snapshot e5e3caa2 saved
Tue Jul 11 15:41:22 PDT 2017
using parent snapshot e5e3caa2
scan [/data/bigstore]
scanned 154557 directories, 3187917 files in 0:07
[33:56] 100.00%  6.297 GiB/s  12.519 TiB / 12.519 TiB  3342551 / 3342474 items  0 errors  ETA 0:00 
duration: 33:56, 6446.62MiB/s
snapshot a34df7d9 saved
Tue Jul 11 22:15:39 PDT 2017
using parent snapshot a34df7d9
scan [/data/bigstore]
scanned 154564 directories, 3187998 files in 0:07
[33:51] 100.00%  6.312 GiB/s  12.520 TiB / 12.520 TiB  3342562 / 3342562 items  0 errors  ETA 0:00 
duration: 33:51, 6462.70MiB/s
snapshot 79fd5fbf saved

Nice speeds.

**Comment by fd0** [comment]
So, now the interesting question: How much time does a `forget` and `prune` operation take when you remove a single snapshot? `prune` is the most costly operation right now...

**Comment by fd0** [comment]
FYI: I'm also experiencing the SSH_FX_FAILURE error and I'm currently debugging it (only occurs after ~30 minutes of runtime with a repo on sftp on a server in my local LAN).

**Comment by fd0** [comment]
Ok, that error was caused by myself in unreleased code. If you ever want to reproduce it, here is how to create a debug log of the sftp server log:
```
$ restic -o sftp.command="ssh SERVER '/usr/lib/sftp-server -e -l DEBUG3'" backup ~/foo 2>/tmp/sftp-debug-log
```

**Comment by todd1794** [comment]
Can we have sftp as a backup source?

Todd

**Comment by fd0** [comment]
Eventually, but not short-term. This was already requested in a different issue (that I can't find right now), but sftp is not a nice protocol and I'm not sure how great that works. Besides, the current archiver code needs to be refactored quite heavily, because it relies on local operations.

**Comment by fd0** [comment]
We've recently merged #1040 which adds a local metadata cache that enormously speeds up operations. I'm therefore closing this issue for now. Please feel free to add further comments. Thanks!
